Add support for NWB schema 2.6.0 and prepare PyNWB 2.3.0

rly opened the pull request:
NWB schema 2.6.0 will be released around January 5, 2023.

Currently, it contains the following [changes](https://github.com/NeurodataWithoutBorders/nwb-schema/blob/dev/docs/format/source/format_release_notes.rst):
- Added OnePhotonSeries.
- ``Subject.age`` has a new optional attribute, ``reference``, which can take a value of "birth" (default) or "gestational".
- Add "in seconds" to the doc of Units.spike_times.

Required PyNWB changes:
- [x] Add support for `Subject.age__reference`: #1540
- [x] Add support for `OnePhotonSeries`: #1593

When merging this branch into `dev`, do NOT squash and merge, so that individual PR contributions are kept. 

This branch should be merged after the NWB schema 2.6.0 release. A PyNWB release (2.2.0) should be made shortly after.

bendichter committed:
add age__reference and adjust tests (#1540)

* add age__reference and adjust tests

* add to CHANGELOG.md

* add period

* Update file.py

* Update CHANGELOG.md

* Update src/pynwb/file.py

* add arg check for Subject. If age__reference is provided, age must also be provided

* add regression test for subject age

* remove raising ValueError when age is omitted

* flake8

* Update CHANGELOG.md

* Update CHANGELOG.md

* Update src/pynwb/file.py

* update to allow Subject.age__reference to be None

* test for Subject.age__reference == None

* Update CHANGELOG.md

* Update CHANGELOG.md

* use mapper to allow None

* update schema

* forbid passing age__reference=None to Subject.__init__

* Update comment

* fix flake8

* Run backwards compat tests in coverage

* Add tests for get_nwb_version

* Fix flake8

* Run IO utils tests in test suite

Co-authored-by: Ryan Ly <rly@lbl.gov>

oruebel commented on src/pynwb/io/utils.py:
```diff
@@ -0,0 +1,27 @@
+    """Get the version of the NWB file from the root of the given builder, as a tuple.
+
+    If the "nwb_version" attribute on the root builder equals "2.5.1", then (2, 5, 1) is returned.
+    If the "nwb_version" attribute on the root builder equals "2.5.1-alpha", then (2, 5, 1) is returned.
```
Why is the ``alpha`` being dropped in the version? It seems strange that "2.5.1" and "2.5.1-alpha" both translate to the same version?

rly replied:
The version tuple is used to branch the API behavior based on the version, e.g., do X if version < 2.5.1, else Y. Version "2.5.1-alpha" should appear only in development. The features in schema "2.5.1-alpha" will be included in "2.5.1" and so for this purpose, the downstream behavior should treat "2.5.1-alpha" and "2.5.1" as equivalent. 

We could refactor this function so that a suffix like "-alpha" is also returned. In Python `(2, 5, 1, "-alpha") > (2, 5, 1)`, so the branching code above should still work.

oruebel replied:
I would suggest to maybe add a flag (e.g., ``include_prerelease=False``) so that a user can configure whether the prerelease tag (alpha, beta etc.) should be included or not.

rly replied:
done

CodyCBakerPhD committed:
Adding OnePhotonSeries class (#1593)

Co-authored-by: Oliver Ruebel <oruebel@users.noreply.github.com>

mavaylon1 committed:
nwb_schema_2.6.0 (#1636)

* Check nwb_version on read (#1612)

* Added NWBHDF5IO.nwb_version property and check for version on NWBHDF5IO.read
* Updated icephys tests to skip version check when writing non NWBFile container
* Add tests for NWB version check on read
* Add unit tests for NWBHDF5IO.nwb_version property
* Updated changelog

Co-authored-by: Ryan Ly <rly@lbl.gov>

* Bump setuptools from 65.4.1 to 65.5.1 (#1614)

Bumps [setuptools](https://github.com/pypa/setuptools) from 65.4.1 to 65.5.1.
- [Release notes](https://github.com/pypa/setuptools/releases)
- [Changelog](https://github.com/pypa/setuptools/blob/main/CHANGES.rst)
- [Commits](https://github.com/pypa/setuptools/compare/v65.4.1...v65.5.1)

---
updated-dependencies:
- dependency-name: setuptools
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* modify export.rst to have proper links to the NWBFile API docs (#1615)

* Create project_action.yml (#1617)

* Create project_action.yml

* Update project_action.yml

* Update project_action.yml

* Update project_action.yml (#1620)

* Update project_action.yml (#1623)

* Project action (#1626)

* Create project_action.yml

* Update project_action.yml

* Update project_action.yml

* Update project_action.yml

* Show recommended usaege for hdf5plugin in tutorial (#1630)

* Show recommended usaege for hdf5plugin in tutorial

* Update docs/gallery/advanced_io/h5dataio.py

* Update docs/gallery/advanced_io/h5dataio.py

Co-authored-by: Heberto Mayorquin <h.mayorquin@gmail.com>

Co-authored-by: Ben Dichter <ben.dichter@gmail.com>
Co-authored-by: Heberto Mayorquin <h.mayorquin@gmail.com>

* Update iterative write and parallel I/O tutorial (#1633)

* Update iterative write tutorial
* Update doc makefiles to clean up files created by the advanced io tutorial
* Fix #1514  Update parallel I/O tutorial to use H5DataIO instead of DataChunkIterator to setup data for parallel write
* Update changelog
* Fix flake8
* Fix broken external links
* Update make.bat
* Update CHANGELOG.md
* Update plot_iterative_write.py
* Update docs/gallery/advanced_io/plot_iterative_write.py

Co-authored-by: Ryan Ly <rly@lbl.gov>

* Update project_action.yml (#1632)

* nwb_schema_2.6.0

* Update CHANGELOG.md

* remove

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Oliver Ruebel <oruebel@users.noreply.github.com>
Co-authored-by: Ryan Ly <rly@lbl.gov>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Ben Dichter <ben.dichter@gmail.com>
Co-authored-by: Heberto Mayorquin <h.mayorquin@gmail.com>

mavaylon1 committed:
Merge branch 'dev' into schema_2.6.0

mavaylon1 committed:
Update Legal.txt

mavaylon1 committed:
Update README.rst

mavaylon1 committed:
Update license.txt

mavaylon1 committed:
Update conf.py

mavaylon1 committed:
Update requirements-min.txt

mavaylon1 committed:
Update requirements.txt

mavaylon1 committed:
Update setup.py

rly committed:
Merge branch 'dev' into schema_2.6.0

rly committed:
Update CHANGELOG.md

rly committed:
Add flag to get_nwb_version to include prerelease info (#1639)

rly committed:
Merge branch 'dev' into schema_2.6.0

rly committed:
Update CHANGELOG.md

rly committed:
Update requirements-min.txt

rly committed:
Update requirements.txt

rly committed:
Update setup.py

rly committed:
Update CHANGELOG.md

rly committed:
Merge branch 'dev' into schema_2.6.0

mavaylon1 committed:
sphinx conf

mavaylon1 commented:
@rly I fixed the sphinx issue but the conda 3.7 from #1427 is now appearing again.

rly committed:
Update CHANGELOG.md

mavaylon1 commented:
Fix #1642

rly committed:
Merge branch 'dev' into schema_2.6.0

rly committed:
Update CHANGELOG.md

rly committed:
Update CHANGELOG.md
